Update "Web renderers" page (#11383)
flutter/website: src/content/platform-integration/web/renderers.md

description: Choosing build modes and renderers for a Flutter web app.
---

Flutter web offers two _build modes_, and two _renderers_.
The two build modes are the **default** or **WebAssembly**,
and the two renderers are **canvaskit** and **skwasm**.

The build mode is chosen when building the app,
and determines which renderers are available at runtime.

> When the default build mode is used,
> the `canvaskit` renderer is used at runtime.
> Alternatively, when the WebAssembly build mode is used,
> Flutter will choose the `skwasm` renderer at runtime,
> and fall back to `canvaskit` if the browser doesn't support `skwasm`.

Reviewer: ```suggestion
For a default build,
Flutter chooses the`canvaskit` renderer at runtime.
For a WebAssembly build,
Flutter chooses the `skwasm` renderer at runtime,
and falls back to `canvaskit` if the browser doesn't support `skwasm`.
```

Re wrote to remove the passive AND future tense. :)

## Build modes

### Default build mode

> The default mode is used when `flutter run` and `flutter build web` commands are
> used without passing `--wasm` or by passing `--no-wasm`.

Reviewer: ```suggestion
Flutter chooses the default mode when the
`flutter run` or `flutter build web` commands are
used without passing `--wasm`, or when passing `--no-wasm`.
```

This build mode only uses the `canvaskit` renderer.

To run in a Chrome using the default build mode:

```console
flutter run -d chrome
```

To build your app for release using the default build mode:

```console
flutter build web
```

### WebAssembly build mode

This mode is enabled by passing `--wasm` to `flutter run` and
`flutter build web` commands.

This mode makes both `skwasm` and `canvaskit` available. `skwasm` requires
[WasmGC][], which is not yet supported by all modern browsers.

> Therefore, at runtime Flutter will pick `skwasm` if garbage collection is
> supported, and fallback to `canvaskit` if not. This allows apps compiled in the
> WebAssembly mode to still run in all modern browsers.

Reviewer: ```suggestion
Therefore, at runtime Flutter chooses `skwasm` if garbage collection is
supported, and falls back to `canvaskit` if not. This allows apps compiled in the
WebAssembly mode to still run in all modern browsers.
```

The `--wasm` flag is not supported by non-web platforms.

To run in Chrome using the WebAssembly mode:

```console
flutter run -d chrome --wasm
```

To build your app for release using the WebAssembly mode:

```console
flutter build web --wasm
```

## Renderers

Flutter has two renderers (`canvaskit` and `skwasm`)
that re-implement the Flutter engine to run the browser. 
The renderer converts UI primitives (stored as `Scene` objects) into
pixels.

### canvaskit

The `canvaskit` renderer is compatible with all modern browsers, and is the 
renderer that is used in the _default_ build mode.

It includes a copy of Skia compiled to WebAssembly, which adds
about 1.5MB in download size.

### skwasm

The `skwasm` renderer is a more compact version of Skia
that is compiled to WebAssembly and supports rendering on a separate thread.

This renderer must be used with the _WebAssembly_ build mode,
which compiles the Dart code to WebAssembly.

To take advantage of multiple threads,
the web server must meet the [SharedArrayBuffer security requirements][].
In this mode,

> Flutter will use a dedicated [web worker][] to offload part of the rendering

Reviewer: ```suggestion
Flutter uses a dedicated [web worker][] to offload part of the rendering
```
